### PR TITLE
Uses class hierarchy to find type and format properties

### LINKED ISSUE #433: Subclassed Lists marked as string type
I am using a tool to generate a Marshmallow schema that I then present as a RESTful API. In order to override some of the methods of marshmallow.fields.List, the generated schema can have generated classes that subclass List. The resulting APISpec is _almost_ perfect because openapi.py mostly uses isinstance to see if the field is a list. For example:

        is_array = isinstance(
            field, (marshmallow.fields.Nested, marshmallow.fields.List)
        )

However, when working out the field type the mapping is looked up using a concrete type:

        self.field_mapping.get(type(field), ("string", None))

It isn't straightforward for me to add the new types using map_to_openapi_type because of the generated nature of my marshmallow and the way the presentation layer is kept separate. However, instead of falling back to `("string", None)`, we can either: fall back to `("array", None)` for instances of a List (and Nested?) either by explicitly checking for those types in `field2type_and_format`; or walking through the field_mapping (in order), checking for instances of each type.

### PR DESCRIPTION
**Pull request by andrjohn**
When sub-classing a marshmallow type then the type and format properties are not found because those properties are looked up in a mapping using concrete types. Other properties are added using `isinstance` to check if the type is a form or `List` or `Nested`.

This change uses the class hierarchy (using the MRO) to do a better job of finding a match in the field mapping. This will reduce the need for custom classes to be added to the field mapping, although that always remains an option if the user wants to control it.

This resolves #433 and #250
This also seems related to #429

### CODE REVIEW

**Comment by andrjohn** [comment]
This PR is made up of a number of commits. I tidy up the tests a bit before adding tests for the new behaviour. It is probably easier to review per commit than all in one go and I can drop commits the maintainers aren't happy with.

**Comment by sloria** [comment]
Thanks for the PR! I'll try to get review this within the next few days.

**Comment by lafrech** [comment]
I didn't get the time to review the main modification in this PR, but I approve the 3 first commits already (the tests rework). Thanks for cleaning this up.

**Comment by dpgaspar** [comment]
Will check if it solves #429 on monday

File: src/apispec/ext/marshmallow/openapi.py
Comment (lafrech) [inline src/apispec/ext/marshmallow/openapi.py]: s/heirarchy/hierarchy

**Comment by lafrech** [comment]
Reviewed. Looks good to me !

File: src/apispec/ext/marshmallow/openapi.py
Comment (sloria) [inline src/apispec/ext/marshmallow/openapi.py]:
```suggestion
```

File: src/apispec/ext/marshmallow/openapi.py
Comment (sloria) [inline src/apispec/ext/marshmallow/openapi.py]:
```suggestion
```

File: src/apispec/ext/marshmallow/openapi.py
Comment (sloria) [inline src/apispec/ext/marshmallow/openapi.py]:
```suggestion
        # hierarchy until we find something that does.
```

File: src/apispec/ext/marshmallow/openapi.py
Comment (sloria) [inline src/apispec/ext/marshmallow/openapi.py]:
```suggestion
```

File: src/apispec/ext/marshmallow/openapi.py
Comment (sloria) [inline src/apispec/ext/marshmallow/openapi.py]: Not exactly sure of the difference, but perhaps `inspect.getmro` is preferred?

**Review by sloria — APPROVED** [review]
Just added a few nits. Since @lafrech has reviewed the substantive changes, this is good to go after the nits are addressed.

File: src/apispec/ext/marshmallow/openapi.py
Comment (lafrech) [reply src/apispec/ext/marshmallow/openapi.py]: Why not

```py
    type(field).__mro__
```

Or even 

```py
    field.__class__.__mro__
```

Is it discouraged to access those attributes?

We do it already:

https://github.com/marshmallow-code/apispec/blob/0845c66b67998f8e47b624c10efebe7bd5dbd532/src/apispec/ext/marshmallow/common.py#L120

Comment (lafrech) [reply src/apispec/ext/marshmallow/openapi.py]: OK, apparently, using `type` is preferred over accessing `__class__` (https://stackoverflow.com/questions/9610993/python-type-or-class-or-is). Unless we're using old style PY2 classes, which shouldn't be the case if inheriting `Field`.

I didn't see any advice against accessing `__mro__` (https://stackoverflow.com/questions/2010692/what-does-mro-do), and I think it is more efficient than calling `mro()` (which does not matter in the context).

Comment (sloria) [reply src/apispec/ext/marshmallow/openapi.py]: Thanks for doing the research @lafrech. So it sounds like ` type(field).__mro__` is preferred. Let's go with it.

Comment (lafrech) [reply src/apispec/ext/marshmallow/openapi.py]: Yeah, no big deal, all of these are functional anyway.

I can change `schema.__class__` into `type(schema)` in common.py someday.

**Comment by andrjohn** [comment]
Thanks @lafrech and @sloria for the quick review. I've pushed a new version with the nits fixed.

**Comment by dpgaspar** [comment]
No it does not solve #429

**Comment by sloria** [comment]
This is released in 1.3.0.

@andrjohn I've invited you to have commit bits. No pressure to do any more work than you already have. Just want to open the door to further collaboration. Feel free to manage issues and review PRs as you fit.

**Comment by lafrech** [comment]
Thanks @andrjohn for your work on this.

I'm realizing perhaps the doc about Custom Fields should be updated (https://github.com/marshmallow-code/apispec/blob/8ed68fb305a302a4691cd04742eb2412b9603718/docs/using_plugins.rst#custom-fields):

- Mention the fact the type of the parent class is used.
- Maybe change the 2nd example as it demonstrates a case that is now covered by default. Just don't let the class inherit from `Integer`.

**Comment by sloria** [comment]
Good catch @lafrech. Issue: https://github.com/marshmallow-code/apispec/issues/437

**Pull request merged**
